File: packages/vitals_domain/lib/src/canonical.dart
import 'clock.dart';
import 'fact.dart';
import 'record.dart';

/// One byte encoding of a record, version 1. Big-endian, length-prefixed, no
/// floats, facts in stamp order — so two devices holding the same set of
/// facts produce identical bytes (invariant 4), whatever order the facts
/// arrived in. A checked-in fixture asserts these bytes; a change to it is a
/// change to what "identical" means for every record in the world.
abstract final class Canonical {
  static const int version = 1;

  static List<int> bytesOf(Record r) {
    final out = <int>[version];
    _bytes(out, r.patient, width: 1);
    final facts = r.all;
    _u32(out, facts.length);
    for (final f in facts) {
      _bytes(out, f.id, width: 1);
      _bytes(out, f.patient, width: 1);
      out.add(f.kind.code);
      _i64(out, f.stamp.wallMillis);
      _u32(out, f.stamp.counter);
      _string(out, f.stamp.device);
      _string(out, f.author);
      _bytes(out, f.payload, width: 4);
      if (f.supersedes == null) {
        out.add(0);
      } else {
        out.add(1);
        _bytes(out, f.supersedes!, width: 1);
      }
    }
    return out;
  }

  static Record recordFrom(List<int> bytes) {
    final r = _Reader(bytes);
    final v = r.u8();
    if (v != version) throw CanonicalError('version $v');
    final patient = r.bytes(width: 1);
    final n = r.u32();
    final facts = <Fact>[];
    for (var i = 0; i < n; i++) {
      final id = r.bytes(width: 1);
      final p = r.bytes(width: 1);
      final kind = FactKind.fromCode(r.u8());
      final wall = r.i64();
      final counter = r.u32();
      final device = r.string();
      final author = r.string();
      final payload = r.bytes(width: 4);
      final hasSup = r.u8();
      final sup = hasSup == 1 ? r.bytes(width: 1) : null;
      facts.add(Fact(
        id: id,
        patient: p,
        kind: kind,
        stamp: Stamp(wallMillis: wall, counter: counter, device: device),
        author: author,
        payload: payload,
        supersedes: sup,
      ));
    }
    if (!r.done) throw CanonicalError('trailing bytes');
    return Record.of(patient, facts);
  }

  static void _u8(List<int> out, int v) => out.add(v & 0xff);
  static void _u32(List<int> out, int v) {
    for (var s = 24; s >= 0; s -= 8) {
      out.add((v >> s) & 0xff);
    }
  }

  static void _i64(List<int> out, int v) {
    for (var s = 56; s >= 0; s -= 8) {
      out.add((v >> s) & 0xff);
    }
  }

  static void _bytes(List<int> out, List<int> b, {required int width}) {
    if (width == 1) {
      if (b.length > 255) throw CanonicalError('too long for one byte');
      _u8(out, b.length);
    } else {
      _u32(out, b.length);
    }
    out.addAll(b);
  }

  static void _string(List<int> out, String s) {
    final units = <int>[];
    for (final c in s.runes) {
      if (c < 0x80) {
        units.add(c);
      } else if (c < 0x800) {
        units.add(0xc0 | (c >> 6));
        units.add(0x80 | (c & 0x3f));
      } else if (c < 0x10000) {
        units.add(0xe0 | (c >> 12));
        units.add(0x80 | ((c >> 6) & 0x3f));
        units.add(0x80 | (c & 0x3f));
      } else {
        units.add(0xf0 | (c >> 18));
        units.add(0x80 | ((c >> 12) & 0x3f));
        units.add(0x80 | ((c >> 6) & 0x3f));
        units.add(0x80 | (c & 0x3f));
      }
    }
    _bytes(out, units, width: 1);
  }
}

final class CanonicalError implements Exception {
  const CanonicalError(this.reason);
  final String reason;
  @override
  String toString() => 'CanonicalError: $reason';
}

final class _Reader {
  _Reader(this._b);
  final List<int> _b;
  var _i = 0;

  bool get done => _i == _b.length;

  int u8() {
    if (_i >= _b.length) throw const CanonicalError('truncated');
    return _b[_i++];
  }

  int u32() {
    var v = 0;
    for (var k = 0; k < 4; k++) {
      v = (v << 8) | u8();
    }
    return v;
  }

  int i64() {
    var v = 0;
    for (var k = 0; k < 8; k++) {
      v = (v << 8) | u8();
    }
    return v.toSigned(64);
  }

  List<int> bytes({required int width}) {
    final n = width == 1 ? u8() : u32();
    if (_i + n > _b.length) throw const CanonicalError('truncated');
    final out = _b.sublist(_i, _i + n);
    _i += n;
    return out;
  }

  String string() {
    final units = bytes(width: 1);
    return String.fromCharCodes(_decodeUtf8(units));
  }

  static List<int> _decodeUtf8(List<int> u) {
    final out = <int>[];
    var i = 0;
    while (i < u.length) {
      final b = u[i];
      if (b < 0x80) {
        out.add(b);
        i += 1;
      } else if (b < 0xe0) {
        out.add(((b & 0x1f) << 6) | (u[i + 1] & 0x3f));
        i += 2;
      } else if (b < 0xf0) {
        out.add(
            ((b & 0x0f) << 12) | ((u[i + 1] & 0x3f) << 6) | (u[i + 2] & 0x3f));
        i += 3;
      } else {
        out.add(((b & 0x07) << 18) |
            ((u[i + 1] & 0x3f) << 12) |
            ((u[i + 2] & 0x3f) << 6) |
            (u[i + 3] & 0x3f));
        i += 4;
      }
    }
    return out;
  }
}

File: server/src/Vitals.Domain/Canonical.cs
using System.Text;

namespace Vitals.Domain;

/// <summary>
/// The byte encoding of a record, version 1 — the same bytes the Dart produces
/// for the same facts, and docs/parity is the proof.
/// </summary>
public static class Canonical
{
    public const byte Version = 1;

    public static byte[] BytesOf(Record r)
    {
        var o = new List<byte> { Version };
        BytesN(o, r.Patient, 1);
        var facts = r.All;
        U32(o, facts.Count);
        foreach (var f in facts)
        {
            BytesN(o, f.Id, 1);
            BytesN(o, f.Patient, 1);
            o.Add((byte)f.Kind);
            I64(o, f.Stamp.WallMillis);
            U32(o, f.Stamp.Counter);
            Str(o, f.Stamp.Device);
            Str(o, f.Author);
            BytesN(o, f.Payload, 4);
            if (f.Supersedes is null) o.Add(0);
            else { o.Add(1); BytesN(o, f.Supersedes, 1); }
        }
        return o.ToArray();
    }

    public static Record RecordFrom(ReadOnlySpan<byte> bytes)
    {
        var r = new Reader(bytes.ToArray());
        var v = r.U8();
        if (v != Version) throw new CanonicalException($"version {v}");
        var patient = r.Bytes(1);
        var n = r.U32();
        var facts = new List<Fact>(n);
        for (var i = 0; i < n; i++)
        {
            var id = r.Bytes(1);
            var p = r.Bytes(1);
            var kind = (FactKind)r.U8();
            var wall = r.I64();
            var counter = r.U32();
            var device = r.Str();
            var author = r.Str();
            var payload = r.Bytes(4);
            var hasSup = r.U8();
            var sup = hasSup == 1 ? r.Bytes(1) : null;
            facts.Add(new Fact(id, p, kind, new Stamp(wall, counter, device), author, payload, sup));
        }
        if (!r.Done) throw new CanonicalException("trailing bytes");
        return Record.Of(patient, facts);
    }

    private static void U32(List<byte> o, int v) { for (var s = 24; s >= 0; s -= 8) o.Add((byte)((v >> s) & 0xff)); }
    private static void I64(List<byte> o, long v) { for (var s = 56; s >= 0; s -= 8) o.Add((byte)((v >> s) & 0xff)); }
    private static void BytesN(List<byte> o, byte[] b, int width)
    {
        if (width == 1) { if (b.Length > 255) throw new CanonicalException("too long for one byte"); o.Add((byte)b.Length); }
        else U32(o, b.Length);
        o.AddRange(b);
    }
    private static void Str(List<byte> o, string s) => BytesN(o, Encoding.UTF8.GetBytes(s), 1);

    private sealed class Reader(byte[] b)
    {
        private int _i;
        public bool Done => _i == b.Length;
        public byte U8() { if (_i >= b.Length) throw new CanonicalException("truncated"); return b[_i++]; }
        public int U32() { var v = 0; for (var k = 0; k < 4; k++) v = (v << 8) | U8(); return v; }
        public long I64() { long v = 0; for (var k = 0; k < 8; k++) v = (v << 8) | U8(); return v; }
        public byte[] Bytes(int width)
        {
            var n = width == 1 ? U8() : U32();
            if (_i + n > b.Length) throw new CanonicalException("truncated");
            var slice = b[_i..(_i + n)];
            _i += n;
            return slice;
        }
        public string Str() => Encoding.UTF8.GetString(Bytes(1));
    }
}

public sealed class CanonicalException(string reason) : Exception(reason);
